driver.go

```diff
@@ -123,32 +123,51 @@ func (d *sqlDriver) unlock() {
 	d.mu.Unlock()
 }
 
-// Open returns a new connection to the database.  The name is a string in a
-// driver-specific format.
+// parseDriverDSN cuts off DB name from dsn. It returns error if the dsn is not
+// valid.
+func parseDriverDSN(dsn string) (storePath, dbName string, err error) {
+	u, err := url.Parse(dsn)
+	if err != nil {
+		return "", "", errors.Trace(err)
+	}
+	path := filepath.Join(u.Host, u.Path)
+	dbName = filepath.Clean(filepath.Base(path))
+	if dbName == "" || dbName == "." || dbName == string(filepath.Separator) {
+		return "", "", errors.Errorf("invalid DB name %q", dbName)
+	}
+	// cut off dbName
+	path = filepath.Clean(filepath.Dir(path))
+	if path == "" || path == "." || path == string(filepath.Separator) {
+		return "", "", errors.Errorf("invalid dsn %q", dsn)
+	}
+	u.Path, u.Host = path, ""
+	return u.String(), dbName, nil
+}
+
+// Open returns a new connection to the database.
+//
+// The dsn must be formatted as an URL 'engine://path/dbname?params'.
```

Comment:
must be a URL format 'engine://path/dbname?params'

btw, don't use an URL, but a URL.